File: packages/sdk-ts/src/voice-session.ts
/**
 * VoiceSession — drive the OpenVox WS pipeline from any TS environment.
 *
 * In Node:
 *
 *   import { VoiceSession } from "@openvox/sdk";
 *   const sess = new VoiceSession({
 *     baseWsUrl: "ws://localhost:3001",
 *     options: { agentId: "abc..." },
 *   });
 *   sess.on(ev => console.log(ev));
 *   await sess.start();
 *   sess.sendPcm(new Int16Array(...));
 *
 * In the browser, use VoiceSession.fromMicrophone() which also handles
 * `getUserMedia` and PCM resampling.
 */

import type { VoiceEvent, VoiceSessionOptions } from "./types.js";

type Listener = (ev: VoiceEvent) => void;

export class VoiceSession {
  private ws: WebSocket | null = null;
  private listeners: Listener[] = [];

  constructor(
    private readonly cfg: { baseWsUrl: string; options?: VoiceSessionOptions; path?: string },
  ) {}

  on(fn: Listener) {
    this.listeners.push(fn);
    return () => {
      this.listeners = this.listeners.filter((f) => f !== fn);
    };
  }

  private emit(ev: VoiceEvent) {
    for (const fn of this.listeners) fn(ev);
  }

  async start(): Promise<void> {
    const url = `${this.cfg.baseWsUrl.replace(/\/$/, "")}${this.cfg.path ?? "/ws/voice"}`;
    const WSImpl = typeof WebSocket !== "undefined" ? WebSocket : ((await import("ws")).default as unknown as typeof WebSocket);
    const ws = new WSImpl(url);
    (ws as any).binaryType = "arraybuffer";
    this.ws = ws;
    await new Promise<void>((res, rej) => {
      ws.onopen = () => res();
      ws.onerror = () => rej(new Error("ws error"));
    });
    const o = this.cfg.options ?? {};
    ws.send(
      JSON.stringify({
        type: "start",
        agent_id: o.agentId,
        system_prompt: o.systemPrompt,
        llm_provider: o.llmProvider,
        llm_model: o.llmModel,
        stt_provider: o.sttProvider,
        tts_provider: o.ttsProvider,
        voice_id: o.voiceId,
        voice_language: o.voiceLanguage,
        sample_rate: o.sampleRate ?? 16000,
      }),
    );

    ws.onmessage = (e: any) => {
      if (typeof e.data === "string") {
        const obj = JSON.parse(e.data);
        this.emit(obj as VoiceEvent);
      } else {
        this.emit({
          type: "audio",
          chunk: e.data as ArrayBuffer,
          sampleRate: 24000,
          encoding: "pcm16",
        });
      }
    };
    ws.onclose = () => {
      this.ws = null;
    };
  }

  sendPcm(pcm: Int16Array | ArrayBuffer): void {
    if (!this.ws) throw new Error("not started");
    const data = pcm instanceof Int16Array ? pcm.buffer : pcm;
    this.ws.send(data as any);
  }

  end(): void {
    if (!this.ws) return;
    this.ws.send(JSON.stringify({ type: "end" }));
  }

  interrupt(): void {
    if (!this.ws) return;
    this.ws.send(JSON.stringify({ type: "interrupt" }));
  }

  close(): void {
    if (this.ws) {
      try {
        this.ws.close();
      } catch {
        /* ignore */
      }
      this.ws = null;
    }
  }
}

File: packages/sdk-py/openvox_sdk/voice_session.py
"""Async client for the OpenVox voice WebSocket pipeline."""

from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Any

import websockets


@dataclass
class VoiceSessionOptions:
    agent_id: str | None = None
    system_prompt: str | None = None
    llm_provider: str | None = None
    llm_model: str | None = None
    stt_provider: str | None = None
    tts_provider: str | None = None
    voice_id: str | None = None
    voice_language: str | None = None
    sample_rate: int = 16000


class VoiceSession:
    """Bidirectional connection to the OpenVox WS pipeline."""

    def __init__(self, base_url: str, options: VoiceSessionOptions | None = None) -> None:
        self._url = base_url.replace("http", "ws", 1).rstrip("/") + "/ws/voice"
        self._opts = options or VoiceSessionOptions()
        self._ws: websockets.WebSocketClientProtocol | None = None

    async def __aenter__(self) -> "VoiceSession":
        await self.start()
        return self

    async def __aexit__(self, *exc: object) -> None:
        await self.close()

    async def start(self) -> None:
        self._ws = await websockets.connect(self._url, max_size=16 * 1024 * 1024)
        await self._ws.send(
            json.dumps(
                {
                    "type": "start",
                    "agent_id": self._opts.agent_id,
                    "system_prompt": self._opts.system_prompt,
                    "llm_provider": self._opts.llm_provider,
                    "llm_model": self._opts.llm_model,
                    "stt_provider": self._opts.stt_provider,
                    "tts_provider": self._opts.tts_provider,
                    "voice_id": self._opts.voice_id,
                    "voice_language": self._opts.voice_language,
                    "sample_rate": self._opts.sample_rate,
                }
            )
        )

    async def send_pcm(self, pcm: bytes) -> None:
        assert self._ws is not None
        await self._ws.send(pcm)

    async def end(self) -> None:
        if self._ws is None:
            return
        await self._ws.send(json.dumps({"type": "end"}))

    async def interrupt(self) -> None:
        if self._ws is None:
            return
        await self._ws.send(json.dumps({"type": "interrupt"}))

    async def events(self) -> AsyncIterator[dict[str, Any]]:
        assert self._ws is not None
        async for raw in self._ws:
            if isinstance(raw, str):
                try:
                    yield json.loads(raw)
                except json.JSONDecodeError:
                    yield {"type": "error", "message": "bad json", "raw": raw}
            else:
                yield {"type": "audio", "chunk": raw}

    async def close(self) -> None:
        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:
                pass
            self._ws = None
